File: scraper.js
const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');

const POSTAL_CODES_FILE = '/home/molt/.openclaw/workspace/postal_codes/postal_codes.txt';
const OUTPUT_DIR = '/home/molt/.openclaw/workspace/postal_codes/data';
const RATE_LIMIT_MS = 1000; // 1 second between requests

// Ensure output directory exists
if (!fs.existsSync(OUTPUT_DIR)) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

// Read postal codes
const postalCodes = fs.readFileSync(POSTAL_CODES_FILE, 'utf-8')
  .split('\n')
  .map(code => code.trim())
  .filter(code => code.length > 0);

console.log(`📍 Starting scraper for ${postalCodes.length} postal codes`);

(async () => {
  let browser;
  let successCount = 0;
  let failureCount = 0;
  const failures = [];

  try {
    browser = await puppeteer.launch({
      headless: true,
      args: ['--no-sandbox', '--disable-setuid-sandbox']
    });

    for (let i = 0; i < postalCodes.length; i++) {
      const postalCode = postalCodes[i];
      console.log(`[${i + 1}/${postalCodes.length}] Processing ${postalCode}...`);

      try {
        const page = await browser.newPage();
        
        // Navigate to the search page with postal code
        const searchUrl = `https://www.posti.fi/postinumerohaku?postinumero=${postalCode}`;
        await page.goto(searchUrl, { waitUntil: 'networkidle2', timeout: 30000 });

        // Wait for results to load
        await page.waitForSelector('.result-item, [class*="result"], [class*="osoite"], table, tr', { timeout: 5000 }).catch(() => {});

        // Extract all data from the page
        const pageData = await page.evaluate((postalCodeParam) => {
          const results = {
            municipalities: new Set(),
            addresses: [],
            pageText: document.body.innerText.substring(0, 5000)
          };

          // Try to find all result items
          const items = document.querySelectorAll('tr, .result-item, [class*="result"], .address, [class*="address"]');
          
          items.forEach(item => {
            const text = (item.textContent || '').trim();
            if (text.length > 2 && text.length < 200) {
              results.addresses.push(text);
              
              // Try to extract municipality names (usually in caps, Finnish/Swedish)
              const munMatch = text.match(/([A-ZÄÖÅ]+)/g);
              if (munMatch) {
                munMatch.forEach(m => {
                  if (m.length > 2 && m.length < 30 && !m.match(/^\d+$/)) {
                    results.municipalities.add(m);
                  }
                });
              }
            }
          });

          return {
            municipalities: Array.from(results.municipalities).slice(0, 10),
            addressCount: results.addresses.length,
            pageText: results.pageText
          };
        }, postalCode);

        // Count unique addresses (removing duplicates)
        let totalAddresses = pageData.addressCount || 0;

        // Try to find count in the page text
        const pageText = pageData.pageText;
        const countMatches = pageText.match(/(\d+)\s*(osoite|address|katu|street)/gi);
        if (countMatches && countMatches.length > 0) {
          const numbers = countMatches.map(m => {
            const num = m.match(/(\d+)/);
            return num ? parseInt(num[1], 10) : 0;
          });
          totalAddresses = Math.max(...numbers) || totalAddresses;
        }

        // Get municipality from page content
        let municipalities = pageData.municipalities || [];
        
        // Try to also get it from the page title or any municipal info
        const pageTitle = await page.evaluate(() => {
          return document.title || '';
        });
        
        if (pageTitle && pageTitle.length > 0) {
          const titleMun = pageTitle.match(/([A-ZÄÖÅ\s]+)/);
          if (titleMun) {
            municipalities.push(titleMun[0].trim());
          }
        }

        // Clean up municipalities
        municipalities = municipalities
          .map(m => m.trim().toUpperCase())
          .filter(m => m.length > 2 && m.length < 50 && m !== postalCode)
          .filter((v, i, a) => a.indexOf(v) === i); // unique

        const result = {
          postal_code: postalCode,
          municipalities: municipalities.slice(0, 5),
          total_addresses: totalAddresses,
          scraped_at: new Date().toISOString()
        };

        // Save to JSON file
        const outputFile = path.join(OUTPUT_DIR, `${postalCode}.json`);
        fs.writeFileSync(outputFile, JSON.stringify(result, null, 2));

        console.log(`✓ ${postalCode}: ${municipalities.join(', ') || 'N/A'} (${totalAddresses} addresses)`);
        successCount++;

        await page.close();

        // Rate limiting
        if (i < postalCodes.length - 1) {
          await new Promise(resolve => setTimeout(resolve, RATE_LIMIT_MS));
        }

      } catch (error) {
        console.error(`✗ ${postalCode}: ${error.message}`);
        failures.push({ postalCode, error: error.message });
        failureCount++;
        
        // Still create a file to mark attempt
        const outputFile = path.join(OUTPUT_DIR, `${postalCode}.json`);
        fs.writeFileSync(outputFile, JSON.stringify({
          postal_code: postalCode,
          municipalities: [],
          total_addresses: 0,
          scraped_at: new Date().toISOString(),
          error: error.message
        }, null, 2));
      }
    }

  } catch (error) {
    console.error('Fatal error:', error);
  } finally {
    if (browser) {
      await browser.close();
    }

    // Print summary
    console.log('\n' + '='.repeat(60));
    console.log('📊 SCRAPING COMPLETE');
    console.log('='.repeat(60));
    console.log(`✓ Successful: ${successCount}/${postalCodes.length}`);
    console.log(`✗ Failed: ${failureCount}/${postalCodes.length}`);

    if (failures.length > 0 && failures.length <= 10) {
      console.log('\n❌ Failures:');
      failures.forEach(f => {
        console.log(`  - ${f.postalCode}: ${f.error}`);
      });
    } else if (failures.length > 10) {
      console.log(`\n❌ ${failures.length} failures (showing first 10):`);
      failures.slice(0, 10).forEach(f => {
        console.log(`  - ${f.postalCode}: ${f.error}`);
      });
    }

    // Sample results
    console.log('\n📋 Sample results (first 5):');
    const dataFiles = fs.readdirSync(OUTPUT_DIR).sort().slice(0, 5);
    dataFiles.forEach(file => {
      if (file.endsWith('.json')) {
        const data = JSON.parse(fs.readFileSync(path.join(OUTPUT_DIR, file), 'utf-8'));
        console.log(`\n${file}:`);
        console.log(JSON.stringify(data, null, 2));
      }
    });

    process.exit(failureCount > postalCodes.length * 0.5 ? 1 : 0);
  }
})();

File: scraper.py
#!/usr/bin/env python3
"""
Posti Postinumerohaku Scraper
Scrapes postal code data from https://www.posti.fi/postinumerohaku
Saves results as JSON files in data/ folder, named by postal code.
"""

import json
import time
import sys
from pathlib import Path
from playwright.sync_api import sync_playwright

# Configuration
POSTAL_CODES_FILE = Path(__file__).parent / "postal_codes.txt"
DATA_DIR = Path(__file__).parent / "data"
POSTI_URL = "https://www.posti.fi/postinumerohaku"

# Ensure data directory exists
DATA_DIR.mkdir(exist_ok=True)

def load_postal_codes():
    """Load postal codes from file."""
    with open(POSTAL_CODES_FILE) as f:
        return [line.strip() for line in f if line.strip()]

def scrape_postal_code(page, postal_code):
    """
    Scrape data for a single postal code.
    Returns dict with postal code, municipalities, and metadata.
    """
    try:
        # Click on search field
        page.fill('[placeholder="Hae osoitteella tai postinumerolla"]', postal_code)
        
        # Wait for dropdown to appear
        page.wait_for_selector('[role="listbox"]', timeout=5000)
        
        # Get all options
        options = page.locator('[role="option"]').all()
        
        results = {
            "postal_code": postal_code,
            "municipalities": [],
            "total_addresses": None,
            "scraped_at": time.strftime("%Y-%m-%d %H:%M:%S")
        }
        
        # Process each municipality option (skip the "Results for..." one)
        for option in options:
            text = option.text_content().strip()
            if text.startswith("Tulokset haulle"):  # Skip result count option
                continue
            if text:
                results["municipalities"].append(text)
        
        # If we have municipalities, click the first one to see total count
        if results["municipalities"]:
            # Click first municipality
            first_option = page.locator('[role="option"]').first
            first_option.click()
            
            # Wait for results table to appear
            page.wait_for_selector('h1:has-text("Hakutulokset")', timeout=5000)
            
            # Extract result count from heading
            heading = page.locator('h1').get_by_text("Hakutulokset").first
            heading_text = heading.text_content()
            
            # Parse "Hakutulokset (Yhteensä XXXX tulosta)" -> extract number
            if "Yhteensä" in heading_text:
                count_str = heading_text.split("Yhteensä")[1].split("tulosta")[0].strip()
                results["total_addresses"] = int(count_str)
        
        return results
    
    except Exception as e:
        return {
            "postal_code": postal_code,
            "error": str(e),
            "scraped_at": time.strftime("%Y-%m-%d %H:%M:%S")
        }

def main():
    """Main scraper loop."""
    postal_codes = load_postal_codes()
    total = len(postal_codes)
    
    print(f"📍 Loading {total} postal codes from {POSTAL_CODES_FILE}")
    print(f"💾 Saving results to {DATA_DIR}/")
    print()
    
    with sync_playwright() as p:
        browser = p.chromium.launch()
        page = browser.new_page()
        page.goto(POSTI_URL)
        
        for i, postal_code in enumerate(postal_codes, 1):
            print(f"[{i:3d}/{total}] Scraping {postal_code}...", end=" ", flush=True)
            
            try:
                # Clear previous search
                page.fill('[placeholder="Hae osoitteella tai postinumerolla"]', "")
                time.sleep(0.5)
                
                # Scrape
                data = scrape_postal_code(page, postal_code)
                
                # Save to JSON file
                output_file = DATA_DIR / f"{postal_code}.json"
                with open(output_file, "w", encoding="utf-8") as f:
                    json.dump(data, f, ensure_ascii=False, indent=2)
                
                if "error" in data:
                    print(f"❌ {data['error']}")
                else:
                    print(f"✅ {len(data['municipalities'])} cities, {data.get('total_addresses', 'N/A')} addresses")
                
                # Rate limiting
                time.sleep(1)
            
            except KeyboardInterrupt:
                print("\n⏹️  Interrupted by user")
                break
            except Exception as e:
                print(f"❌ Exception: {e}")
        
        browser.close()
    
    print()
    print(f"✅ Scraping complete! Results saved to {DATA_DIR}/")

if __name__ == "__main__":
    main()
